### PR TITLE
Wip #16841 uploaded file directive

### PR DESCRIPTION
**Pull request by johanandren**
A new very basic directive for accepting files through multipart uploads. Streams the file onto a user specified location on disk.

Refs #16841

### CODE REVIEW

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  import BasicDirectives._
+  import FutureDirectives._
+  import MarshallingDirectives._
+
+  /**
+   * Collects the bytes of the first file part from the request into the given destination file
+   *
+   * @param destination Will be overwritten if not empty
+   */
+  def uploadedFile(destination: File): Directive1[Try[FormFile]] =
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: How about a more flexible
```scala
def upload(maxFiles: Int = 1): Directive1[Source[(FileInfo, Source[ByteString])]]
```
?

We could then build shortcuts like the above around that.

related: #17665

Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: So, that wouldn't actually store anything on disk, just collect the metadata and the stream for each of the parts?

Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: Yeah, maximum flexibility! ;) That could serve as a base for higher level directives that write files.

Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: Interesting idea @2beaucoup! I wonder if both would be useful? This being `uploadTo(file) { f => }` then?

// cc @sirthias @jrudolph

Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: On one hand I like that, one super simple and one very flexible but on the other hand, that is not very far off from what we already have - basically a filter and a map on the existing stream of body parts? So, bigger api surface for little gain? 

I think the super-simple one is great to keep the threshold for newcomers low and make "the simple things very easy", so we should definitely keep it or something close to it.

Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: My suggestion was to keep both, with the simple one being based on the flexible one. ;)

Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: > My suggestion was to keep both, with the simple one being based on the flexible one. ;)

:+1: 

And in docs we should mention how to get the `dataBytes`, and hint at "you can get the request body as a stream, isn't that amazing" ;)

**Comment by johanandren** [comment]
Some more doc fixes to do, but wanted to push working code to see if there is any more api-feedback first.

**Comment by johanandren** [comment]
PLS BUILD

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  import BasicDirectives._
+  import FutureDirectives._
+  import MarshallingDirectives._
+
+  /**
+   * Collects the bytes of the first file part from the request into the given destination file
+   *
+   * @param destination Will be overwritten if not empty
+   */
+  def uploadedFile(destination: File): Directive1[Try[UploadedFile]] =
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: I'm starting to think it is a bit weird to not talk about field name here, if we look at the Play API for the same thing for example that is kind of central: https://www.playframework.com/documentation/2.0.2/JavaFileUpload

Maybe it should be `uploadedFile(fieldName: String, destination: File)` rather?

Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: makes sense!

**Comment by 2beaucoup** [comment]
Looks like some of my comments were force-pushed away. ;(

`ctx` in `uploadedFiles` is unused and `uploadedFile` should extract a plain `File` so we can get rid of `UploadedFile`.

**Comment by johanandren** [comment]
Yeah, had to rebase on top of the Source.file changes, sorry.

About the using regular `File` instead of `UploadedFile`: The reason I added `UploadedFile` was to keep metadata such as original file name and mime type of the file, which you would otherwise loose when choosing your own file (and which probably aren't entirely safe to just trust but will require some further processing in actual apps)

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        onComplete(uploaded.runWith(Sink.head[UploadedFile]))
+      }
+    }
+
+  /**
+   * Collects each body part that is a multipart file as a tuple containing metadata and a `Source`
+   * for streaming the file contents somewhere.
+   * @param maxFiles Accept at most this many uploaded files, any additional files are ignored
+   */
+  def uploadedFiles(maxFiles: Int): Directive1[Source[(FileInfo, Source[ByteString, Any]), Any]] =
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: What about
```scala 
def uploadedFiles(fieldNames: String*)
``` 
?

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala (hunk 2 above)
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: We could expect the destination to be a folder and reuse the `part.fileName` as the destination file. Another option would be:
```scala
def upload(fieldNames: String*, destination: FileInfo => File): Directive1[Map[FileInfo, Try[File]]]
```

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala (hunk 3 above)
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: :+1:

**Comment by 2beaucoup** [comment]
What we should keep in mind is that when the `multiple` attribute is used in multpart forms (for selecting multiple files at once for one input field) then there are multiple parts with the same `name` but different `filename`s.

**Comment by johanandren** [comment]
I think we should let the API look like it does in the last commit. I'm thinking the following use cases:

1. you are new to the api and just want to accept a simple file upload to your server - `uploadedFile()` is super simple and provides you with all the metadata you need
2. you want to do a more complex form where it is possible to upload multiple files (possibly with the `multiple` attribute) - `uploadedFiles()` then is for you, this also allows streaming the contents somewhere else etc.
3. You want to do something even more complex - you will have to use the "raw" multipart handling, but you can look at the sources of `FileUploadDirectives.scala` to get inspiration.

**Comment by 2beaucoup** [comment]
I'd vote for at least being explicit in what param name should be used instead of just using the first one.

Removing the extraction and just rejecting if something went wrong would simplify things further.

The current interface makes it hard to compute the `destination` from the uploaded `part.filename` which might be a requirement even for the simple case.

I'd therefore vote for:

```scala
// simple:
def uploadFile(fieldName: String, destinationFolder: File): Directive0 // using part.filename
// or
def uploadFile(fieldName: String, destination: FileInfo => File): Directive0

// advanced:
def uploadFiles(fieldNames: String*): Directive1[Source[(FileInfo, Source[ByteString, Any]), Any]]
```

Maybe @sirthias and @jrudolph have something to add here?

**Comment by johanandren** [comment]
How about three different levels of it, all including the field names:
```Scala
// super-simple use case: creates a tmp-file with the contents, 
// if there is a failure the file is removed to not leak disk space
// rejects on error or if there is no such field - this is on par with what play, rails and php provides
def uploadedFile(fieldName: String): Directive1[UploadedFile]

// semi-advanced: 
// I still think you want the metadata extracted, if we do not extract it the user would be 
// encouraged to side-effect-saving the file info to get access to it
// rejects on error or if there is no such field
def uploadedFile(fieldName: String, destination: FileInfo => File): Directive1[UploadedFile]

// advanced use case but still filter on field names, should allow most use cases
// but you always know what fields you have in your forms in advance
// also, all error handling is left to the user
def uploadedFiles(fieldNames: String*): Directive1[Source[(FileInfo, Source[ByteString, Unit], Unit]
```

**Comment by johanandren** [comment]
Or maybe complete with BadRequest/InternalServerError is better than rejecting.

**Comment by ktoso** [comment]
@2beaucoup @johanandren 

As for the `uploadFile`: 
- Rejecting is good – that's what `MissingFormFieldRejection` is for. If users want to 500 they can from handling that rejection.
- `(fieldName: String, destinationFolder: File) // using part.filename` the destinationFolder seems rather meh; wouldn't this encourage usage in which I can easily end up with a corrupt file? Since multiple requests, could concurrently write to the same file - if we base the name on part.filename. I'd vote against using the filename from the req, as real filename on the disk. Thus, we need to provide the user UploadedFile with metadata about the file ("real name").

As for streams:
- getting a stream must be as simple as doing the `request > File`, since we do want to encourage "streams all the way through"
- most likely different name for this would be better; how about `fileUpload("formFieldName") { source =>`, since "fileUpload" sounds like "we'll be uploading a file" and "uploadedFile" is past tense so it implies "there is this file, and you can read it".
  - let's not do `fileUpload[s]` for now (stream of streams)

**Comment by 2beaucoup** [comment]
```scala
def fileUpload(fieldName: String, tempFolder: File): Directive1[(FileInfo, File)]
def fileUpload(fieldName: String, destination: FileInfo => File): Directive1[FileInfo]
def fileUpload(fieldName: String): Directive1[(FileInfo, Source[ByteString, Any])]
```
?

**Comment by johanandren** [comment]
Yeah, after discussion on gitter it will be 
```
def uploadedFile(fieldName: String): Directive1[UploadedFile]
def fileUpload(fieldName: String): Directive1[(FileInfo, Source[ByteString, Any])]
```

for now, and we'll see if there is much request for more advanced ones and add those later

**Comment by 2beaucoup** [comment]
> def uploadedFile(fieldName: String): Directive1[UploadedFile]

But *where* is it stored?

**Comment by johanandren** [comment]
`File.createTempFile`

**Comment by 2beaucoup** [comment]
```java
public static File createTempFile(String prefix, String suffix)
```

or

```java
public static File createTempFile(String prefix, String suffix, File directory)
```

? ;)

A hard coded destination might a bit too inflexible, but should work for now.

**Comment by johanandren** [comment]
The one with the default system tmp file, prefixing the files with "akka-http-upload" and suffixing with ".tmp"

**Comment by ktoso** [comment]
> A hard coded destination might a bit too inflexible, but should work for now.

Let's follow up on that as separate ticket and see if there's user demand for it :)
In this PR let's do the simplest possible thing, the signatures mentioned in: https://github.com/akka/akka/pull/18946#issuecomment-157685268 from there on we're free to add more afterwards (adding is simple, removing is not :wink:)

**Comment by 2beaucoup** [comment]
@johanandren, @ktoso agreed.

One last (promised!) thought: Why not avoid the introduction of `UploadedFile`? `uploadedFile` could extract a `(FileInfo, File)` tuple (analog to `fileUpload` extracting a `(FileInfo, Source[ByteString, Any])` tuple). The size would be accessable via the `File`.

Over and out. :)

**Comment by ktoso** [comment]
> One last (promised!) thought: Why not avoid the introduction of UploadedFile? uploadedFile could extract a (FileInfo, File) tuple (analog to fileUpload extracting a (FileInfo, Source[ByteString, Any]) tuple). The size would be accessable via the File.

I agree :+1:

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+trait FileUploadDirectives {
+
+  import BasicDirectives._
+  import RouteDirectives._
+  import FutureDirectives._
+  import MarshallingDirectives._
+
+  /**
+   * Streams the bytes of the file submitted using multipart with the given file name into a temporary file on disk.
+   * If there is an error writing to disk the request will be rejected, if there is no such field the request will
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: Last nitpick: this is not implemented as rejecting, it uses `failWith` which is "will be failed with the cause exception".
Reject is "this route does not match".

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/file-upload-directives/fileUpload.rst [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+Signature
+---------
+
+.. includecode2:: /../../akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala
+   :snippet: fileUpload
+
+Description
+-----------
+Simple access to the stream of bytes for a file uploaded as a multipart form together with metadata
+about the upload as extracted value.
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/file-upload-directives/fileUpload.rst]: Paste the part about failure handling here

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/file-upload-directives/uploadedFile.rst [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+Signature
+---------
+
+.. includecode2:: /../../akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala
+   :snippet: uploadedFile
+
+Description
+-----------
+Streams the contents of a file uploaded as a multipart form into a temporary file on disk and provides the file and
+metadata about the upload as extracted value.
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/file-upload-directives/uploadedFile.rst]: Paste the part about error handling here

Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/file-upload-directives/uploadedFile.rst]: Add a 
```
.. _note: 
  This directive will stream contents of the request into a file, however one can not start processing these until the file has been written completely. For streaming APIs it is prefered to use the :ref:`-fileUpload-` directive, as it allows for streaming handling of the incoming data bytes.
```

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      case Some(tuple) ⇒ provide(tuple)
+      case None        ⇒ reject(MissingFormFieldRejection(fieldName))
+    }
+
+}
+
+object FileUploadDirectives extends FileUploadDirectives
+
+/**
+ * @param fieldName Name of the form field the file was uploaded in
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: Needs a comment describing when it's passed out, not just the params I'd say

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  import FutureDirectives._
+  import MarshallingDirectives._
+
+  /**
+   * Streams the bytes of the file submitted using multipart with the given file name into a temporary file on disk.
+   * If there is an error writing to disk the request will be rejected, if there is no such field the request will
+   * be rejected, if there are multiple file parts with the same name, the first one will be used and the subsequent
+   * ones ignored.
+   */
+  def uploadedFile(fieldName: String): Directive1[(FileInfo, File)] =
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: these should also work with symbols, to be consistent with the rest of HTTP.
I'll provide a sample how to do that

Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala]: Ignore that :) The `FileAndResourceDirectives` are not defined using symbol/string anyway, only paths are so we're OK here.

**Comment by ktoso** [comment]
LGTM, found last minor things and let's merge then

**Comment by ktoso** [comment]
Thanks!

**Pull request merged**